### PR TITLE
[IntegratedTest] Add test generator

### PR DESCRIPTION
**Pull request by zhoonit**
Add test generator and model recorder for integrated test cases.

This runs multiple epochs and save related informations.

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

File: test/input_gen/genModelTests.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: Apache-2.0
+##
+# Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+#
+# @file recorder.py
+# @brief Generate tc using KerasRecorder
```
Comment (kparichay) [inline test/input_gen/genModelTests.py:7]: missing date?

File: test/input_gen/recorder.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+class KerasRecorder:
+    def __init__(
+        self,
+        file_name,
+        inputs,
+        outputs,
+        input_shape,
+        label_shape,
+        loss_fn=None,
+        optimizer=tf.keras.optimizers.SGD(lr=1.0),
```
Comment (kparichay) [inline test/input_gen/recorder.py:46]: Maybe in a later PR, but you would want another argument here `from_logits`.
In this case, the activation for the last layer (for eg softmax) is amalgamated in the loss itself.
So, the output of the last layer must add that activation by itself (just for returning and not for training) to match the output with nntrainer.
This is needed because amalgamated losses are mathematically more stable than the separated versions.

File: test/input_gen/recorder.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+# @brief Record Keras model with some watchers attached
+# @note  The class might need to go through some rework for non-sequential model
+# in case of the order of graph traversal is diffrent from NNTrainer
+class KerasRecorder:
+    def __init__(
+        self,
+        file_name,
+        inputs,
+        outputs,
+        input_shape,
```
Comment (kparichay) [inline test/input_gen/recorder.py:43]: I think you will need a batch size separately in this

File: test/input_gen/recorder.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        label_shape,
+        loss_fn=None,
+        optimizer=tf.keras.optimizers.SGD(lr=1.0),
+    ):
+        self.inputs = inputs
+        self.outputs = outputs
+        self.model = K.Model(inputs=inputs, outputs=outputs)
+        self.loss_fn = loss_fn
+        self.optimizer = optimizer
+        self.file = open(file_name, "wb")
```
Comment (kparichay) [inline test/input_gen/recorder.py:53]: how about raise if file already exists, or warn?

File: test/input_gen/recorder.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            )
+
+            self.optimizer.apply_gradients(zip(gradients, layer.trainable_weights))
+
+        self._write_items(results[-1])
+
+    ##
+    # @brief run function
+    # @param epoch number of iteration to run
+    def run(self, epoch):
```
Comment (kparichay) [inline test/input_gen/recorder.py:125]: I think this might be better named as iteration

File: test/input_gen/recorder.py (hunk 2 above)
Comment (zhoonit) [reply test/input_gen/recorder.py:46]: Isn't this done by adding 

`loss_fn=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)`
and omit softmax activation on he last layer?

File: test/input_gen/recorder.py (hunk 3 above)
Comment (zhoonit) [reply test/input_gen/recorder.py:43]: Could you explain more about this? I am not getting the benefit from separating the batchsize

File: test/input_gen/genModelTests.py (hunk 1 above)
Comment (zhoonit) [reply test/input_gen/genModelTests.py:7]: updated

File: test/input_gen/recorder.py (hunk 5 above)
Comment (zhoonit) [reply test/input_gen/recorder.py:125]: Done

File: test/input_gen/recorder.py (hunk 4 above)
Comment (zhoonit) [reply test/input_gen/recorder.py:53]: Done

File: test/input_gen/recorder.py (hunk 2 above)
Comment (kparichay) [reply test/input_gen/recorder.py:47]: Yes, this is done for training, but inference, softmax is still needed (which requires extra conditions than just setting `from_logits=True`).

File: test/input_gen/recorder.py (hunk 3 above)
Comment (kparichay) [reply test/input_gen/recorder.py:44]: In tf.keras, input layer takes input_shape and batch_size separately.
As some tests will want to run over different batches, batch_size might be needed.

Comment (zhoonit) [reply test/input_gen/recorder.py:44]: Sounds like it is better to restructure the class to pass input_shape, label_shape and filename to `self.run` phase. To make the behavior consistent, I'll add a test using this python script first and make another PR to restructure this file to make sure things don't break down.

File: test/input_gen/recorder.py (hunk 2 above)
Comment (zhoonit) [reply test/input_gen/recorder.py:47]: Ditto. I will do this in diffrent PR. For now, I'll leave a note in my issue

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM
